apps/gdal_rasterize_lib.cpp

```diff
@@ -1072,6 +1072,82 @@ GDALDatasetH GDALRasterize(const char *pszDest, GDALDatasetH hDstDS,
         }
     }
 
+    const int nLayerCount =
+        (psOptions->osSQL.empty() && psOptions->aosLayers.empty())
+            ? 1
+            : static_cast<int>(psOptions->aosLayers.size());
+
+    const bool bOneSizeNeedsCalculation{
+        static_cast<bool>((psOptions->nXSize == 0) ^ (psOptions->nYSize == 0))};
+
+    // Calculate the size if either nXSize or nYSize is 0
+    if (psOptions->osSQL.empty() && bOneSizeNeedsCalculation)
+    {
+        CPLErr eErr = CE_None;
+        // Get the extent of the source dataset
+        OGREnvelope sEnvelop;
+        bool bFirstLayer = true;
+        for (int i = 0; i < nLayerCount; i++)
+        {
+            OGRLayerH hLayer;
+            if (psOptions->aosLayers.size() > static_cast<size_t>(i))
+                hLayer = GDALDatasetGetLayerByName(
+                    hSrcDataset, psOptions->aosLayers[i].c_str());
+            else
+                hLayer = GDALDatasetGetLayer(hSrcDataset, 0);
+            if (hLayer == nullptr)
+            {
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "Unable to find layer \"%s\".",
+                         psOptions->aosLayers.size() > static_cast<size_t>(i)
+                             ? psOptions->aosLayers[i].c_str()
+                             : "0");
+                eErr = CE_Failure;
+                break;
+            }
+            OGREnvelope sLayerEnvelop;
+            if (OGR_L_GetExtent(hLayer, &sLayerEnvelop, TRUE) != OGRERR_NONE)
+            {
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "Cannot get layer extent");
+                eErr = CE_Failure;
+                break;
+            }
+            if (bFirstLayer)
+            {
+                sEnvelop = sLayerEnvelop;
+                bFirstLayer = false;
+            }
+            else
+            {
+                sEnvelop.Merge(sLayerEnvelop);
+            }
+        }
+        if (eErr == CE_Failure)
+        {
+            if (bCloseOutDSOnError && hDstDS)
+            {
+                GDALClose(hDstDS);
+            }
+            return nullptr;
+        }
+
+        if (psOptions->nXSize == 0)
+        {
+            const_cast<GDALRasterizeOptions *>(psOptions)->nXSize =
+                static_cast<int>(
+                    (sEnvelop.MaxX - sEnvelop.MinX) /
+                    ((sEnvelop.MaxY - sEnvelop.MinY) / psOptions->nYSize));
```

Comment:
I suppose that somewhere in the code the case with a single point input dataset with Envelop.MaxX=Envelop.MinX is considered. Or a case with a single line that is either vertical or horizontal.